dts/bindings/mtd/jedec,spi-nor.yaml

```diff
@@ -24,3 +24,25 @@ properties:
     type: phandle-array
     required: false
     description: RESETn pin
+  page-size:
+    type: int
+    required: false
+    default: 256
+    description: |
+      Page size in bytes.
+  erase-types:
+    type: array
+    required: false
+    default: [0x10000, 0xD8, 0x8000, 0x52, 0x1000, 0x20, 0, 0]
```

Comment:
Remove the default here, an add this to the various dts files that need it.

Reply:
Do we really expect any opcodes beyond 'sector erase' and 'block erase SIZE'?

If so, maybe the encoding for erase-types should be reworked since the opcode for block implies the size.

Reply:
> Do we really expect any opcodes beyond 'sector erase' and 'block erase SIZE'?

Yes

> If so, maybe the encoding for erase-types should be reworked since the opcode for block implies the size.

It doesn't.  We've recently learned that the `cypress,s25fl512s` uses command 0xD8 to erase a 256 KiBy sector---and that's the only erase command it supports.  So we can't infer erase size from the command, and we can't assume any given command is supported.  Page size on that device is still 256 bytes.

I do think that we can benefit from a default for the three sizes you mention above, and only require them to be explicit when the device supports additional or incompatible commands.

Honestly this is quite a mess.  The only fully portable solution is reading the SFDP at runtime, and that's a waste of code if you know what chips your application has to deal with.

Reply:
I choose to add the defaults because this reflects the old behavior of the driver. Currently every in tree spi-nor device works with these defaults. Furthermore adding defaults preserves compatibility for most out of tree uses of this driver (except for users of has_be32k).

About the erase sizes I can only add the following to pabigots answer: JEDS216 allows up to four erase sizes (must be a power of two) with arbitrary op-codes. I wanted to fully support this.

Reply:
> About the erase sizes I can only add the following to pabigots answer: JEDS216 allows up to four erase sizes (must be a power of two) with arbitrary op-codes. I wanted to fully support this.

Can you cite the section that requires the sizes to be a power of two?  One of the things I'm not happy about with the current approach is the wasted space representing the erase size, even though it's reduced since the size and command are in separate arrays.

If we can always express a supported erase size as the power-of-2 exponent, then the `erase-types` value could be a sequence of byte values, e.g.:

    erase-types = <0 0xc7 16 0xd8 12 0x20>; /* CE, SE, BE */

using 0 to indicate "whole chip" and so eliminating the special case for that command.  We shouldn't need to require the trailing zero values.

Reply:
Please see DWORD 8 and 9, page 15 and 16, [JESD216 rev B](https://www.taterli.com/wp-content/uploads/2017/07/JESD216.pdf). Revision B calls erase types sector types. With newer revision of the standard the terminology changed to erase types.

Reply:
Thanks; that's in section 6.4 of JESD216D.  It seems to be consistent with the byte size encoding I propose.

Following that description should we relax the requirement that the pairs be in any specific order?  It'd be pretty easy to scan and identify the one that's best for the requested erase span.

Reply:
In fact, we could simplify and prepare for eventual SFDP support by using JESD216 field names and content for the devicetree properties:
* `sfdp-basic-8` corresponds to the 8th DWORD of the JEDEC Basic Flash Parameter Table, providing erase type 1 and 2 size and instruction.
* `sfdp-basic-9` corresponds to the 9th DWORD of the JEDEC Basic Flash Parameter Table, providing erase type 3 and 4 size and instruction.

AFAICT JESD216B doesn't provide a configurable instruction for chip erase, so the stacked flash memories like MT25QU01GBBB still need a property to indicate that some magic technique is required to erase the entire memory.

Reply:
I like that idea. I guess we should also use this for page size (11th DWORD) and any other future property that is part of SFDP. If you agree, I will rework the driver.
I am not exactly sure how to handle stacked flash memories. From the datasheet of MT25QU01GBBB I conclude that there is no big difference to single die chips except the missing chip erase and the different command for die erase. But JESD216D seems to provide some offsets for control and status registers for multi die devices (Section 6.11) in which case things are not quite as easy.

Maybe a property `die-erase-cmd` and `die-count` would be enough to handle our cases (support for MT25QU01GBBB like chips). In case of a single die, die-count would be 1 and die-erase-cmd set to 0xC7 (chip erase). If die-count > 1 it would be handled like a block erase command. I currently do not have a board with a multi die flash memory so i can not test any features regarding them.

Reply:
I think we're going down a slippery slope here, as there are existing properties related to deep power down timing that could be inferred (with some complication) from `sfdp-basic-12` and `sfdp-basic-13`.  Also parts of `sfdp-basic-11` (e.g. timing) are likely to be ignored.

But even if we don't convert everything right away, I do think switching to SFDP-based specification makes a lot of sense, so let's go that way for anything new, like this.

I'm not really clear on how die erase is supposed to be used on a stacked device.  I also don't have one, nor are the technical notes very informative.